domain/test/unit_cuda/primitives/warpscan.cu

```diff
@@ -23,176 +38,410 @@
 
 using namespace cstone;
 
-__global__ void testMin(int* values)
+__device__ unsigned globalIndex()
 {
-    int laneValue = threadIdx.x;
-
-    values[threadIdx.x] = warpMin(laneValue);
+    const auto blockIndex = blockIdx.x + blockIdx.y * gridDim.x + blockIdx.z * gridDim.x * gridDim.y;
+    return blockIndex * blockDim.x * blockDim.y * blockDim.z + threadIdx.x + threadIdx.y * blockDim.x +
+           threadIdx.z * blockDim.x * blockDim.y;
 }
 
-TEST(WarpScan, min)
+template<class InputT, class OutputT = InputT, class F>
+__global__ void applyWarpFunction(InputT* const input, OutputT* output, F f)
 {
-    thrust::host_vector<int> h_v(GpuConfig::warpSize);
-    thrust::device_vector<int> d_v = h_v;
+    const unsigned index = globalIndex();
+    output[index]        = f(input[index]);
+}
 
-    testMin<<<1, GpuConfig::warpSize>>>(thrust::raw_pointer_cast(d_v.data()));
+struct SomeStruct
+{
+    int a;
+    float b;
+    double c;
+    bool d;
 
-    h_v = d_v;
-    thrust::host_vector<int> reference(GpuConfig::warpSize, 0);
+    bool operator==(SomeStruct const& other) const
+    {
+        return a == other.a && b == other.b && c == other.c && d == other.d;
+    }
+};
 
-    EXPECT_EQ(h_v, reference);
+std::ostream& operator<<(std::ostream& out, SomeStruct const& s)
+{
+    out << "SomeStruct {" << s.a << ", " << s.b << ", " << s.c << ", " << s.d << "}";
+    return out;
 }
 
-__global__ void testMax(int* values)
+template<class T>
+std::tuple<dim3, dim3, thrust::host_vector<T>> testData()
 {
-    int laneValue = threadIdx.x;
+    const dim3 numBlocks = {5, 2, 3};
+    const dim3 blockSize = {GpuConfig::warpSize / 4, 2, 6};
```

Comment:
I think the `dim3` grid and blocks are a distraction that don't contribute to the test. Both the GPU-kernel test data computation and the reference computation operate on a 1D linear sequence.
Other than testing that `globalIndex` correctly enumerates that sequence you could use plain 1D grids/blocks -- which if you really wanted to you could make a separate test for, but I don't think it's necessary.

Reply:
This is meant to test proper lane indexing on 3D blocks, not data indexing. Try changing the `laneIndex()` function to its previous definition of `threadIdx.x & (GpuConfig::warpSize - 1)` and you’ll see that most tests fail. I’ll add a comment to make that clear.